Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift

```diff
@@ -0,0 +1,111 @@
+
+class MockSocksServer {
+    
+    let channel: Channel
+    
+    public init(expectedURL: String, expectedResponse: String, file: String = (#file), line: UInt = #line) throws {
+        let elg = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+        let bootstrap = ServerBootstrap.init(group: elg)
+            .serverChannelOption(ChannelOptions.socket(SocketOptionLevel(SOL_SOCKET), SO_REUSEADDR), value: 1)
+            .childChannelInitializer { channel in
+            let handshakeHandler = SOCKSServerHandshakeHandler()
+            return channel.pipeline.addHandlers([
+                handshakeHandler,
+                SOCKSTestHandler(handshakeHandler: handshakeHandler),
+                SOCKSTestHTTPClient(expectedURL: expectedURL, expectedResponse: expectedResponse, file: file, line: line)
+            ])
+        }
+        self.channel = try bootstrap.bind(host: "127.0.0.1", port: 1080).wait()
+    }
+    
+    func shutdown() throws {
+        try self.channel.close().wait()
+    }
+    
+}
+
+class SOCKSTestHTTPClient: ChannelInboundHandler {
+    
+    typealias InboundIn = HTTPServerRequestPart
+    typealias OutboundOut = HTTPServerResponsePart
+    
+    let expectedURL: String
+    let expectedResponse: String
+    let file: String
+    let line: UInt
+    
+    init(expectedURL: String, expectedResponse: String, file: String, line: UInt) {
+        self.expectedURL = expectedURL
+        self.expectedResponse = expectedResponse
+        self.file = file
+        self.line = line
+    }
+    
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        let message = self.unwrapInboundIn(data)
+        switch message {
+        case .head(let head):
+            XCTAssertEqual(head.uri, self.expectedURL)
+        case .body:
+            break
+        case .end:
+            context.write(self.wrapOutboundOut(.head(.init(version: .http1_1, status: .ok))), promise: nil)
+            context.write(self.wrapOutboundOut(.body(.byteBuffer(context.channel.allocator.buffer(string: self.expectedResponse)))), promise: nil)
+            context.writeAndFlush(self.wrapOutboundOut(.end(nil)), promise: nil)
+        }
+    }
+}
+
+class SOCKSTestHandler: ChannelInboundHandler, RemovableChannelHandler {
+    
+    typealias InboundIn = ClientMessage
+    
+    let handshakeHandler: SOCKSServerHandshakeHandler
+    
+    init(handshakeHandler: SOCKSServerHandshakeHandler) {
+        self.handshakeHandler = handshakeHandler
+    }
+    
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        let message = self.unwrapInboundIn(data)
+        switch message {
+        case .greeting:
+            context.writeAndFlush(.init(
+                ServerMessage.selectedAuthenticationMethod(.init(method: .noneRequired))), promise: nil)
+            context.writeAndFlush(.init(
+                ServerMessage.authenticationData(context.channel.allocator.buffer(capacity: 0), complete: true)), promise: nil)
```

Comment:
Do we need to send authentication data here?

Reply:
We need to send an empty buffer with the `complete` flag set to true so that the server state machine is updated.

Reply:
Oh. Huh. I think the server should have a fast-path for `noneRequired` that can _tolerate_ receiving an empty buffer with `complete`, but will also happily progress without it.

Reply:
Sure, I'll file an issue in swift-nio-extras, or do you think we should provide this fast path *now*?

Reply:
I think we should provide it before we ship swift-nio-extras.

Reply:
Alright

Reply:
https://github.com/apple/swift-nio-extras/pull/134